### PR TITLE
Add support for some FPKI certificate cases, UUID, FASC-N, PIV extension

### PR DESCRIPTION
**Pull request by JacobBarthelmeh**
# Description

Add support for some FPKI certificate cases, UUID, FASC-N, PIV extension

# Testing

Tested with wolfSSH connection and with additional unit test cases added in this PR.

# Checklist

 - [X] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

File: tests/api.c [hunk 1]
```diff
@@ -2260,6 +2260,44 @@ static void test_wolfSSL_CertManagerNameConstraint5(void)
 #endif
 }
 
+static void test_wolfSSL_FPKI(void)
+{
+#if defined(WOLFSSL_FPKI)
```
Comment (dgarske) [inline tests/api.c:2265]: Please wrap with `!defined(NO_FILESYSTEM)`.

File: tests/api.c [hunk 2]
```diff
@@ -2260,6 +2260,44 @@ static void test_wolfSSL_CertManagerNameConstraint5(void)
...
+    printf(testingFmt, "test_wolfSSL_FPKI");
+    f = XFOPEN(fpkiCert, "rb");
+    AssertTrue((f != XBADFILE));
+    bytes = (int)XFREAD(buf, 1, sizeof(buf), f);
+    XFCLOSE(f);
+
+    wc_InitDecodedCert(&cert, buf, bytes, NULL);
+    AssertIntEQ(wc_ParseCert(&cert, CERT_TYPE, 0, NULL), 0);
+    AssertIntEQ(wc_GetFASCNFromCert(&cert, NULL, &fascnSz), LENGTH_ONLY_E) ;
+    fascn = (byte*)XMALLOC(fascnSz, DYNAMIC_TYPE_TMP_BUFFER, NULL);
```
Comment (dgarske) [inline tests/api.c:2285]: Should be `fascn = (byte*)XMALLOC(fascnSz, NULL, DYNAMIC_TYPE_TMP_BUFFER);`

File: tests/api.c [hunk 3]
```diff
@@ -2260,6 +2260,44 @@ static void test_wolfSSL_CertManagerNameConstraint5(void)
...
+    AssertNotNull(fascn);
+    AssertIntEQ(wc_GetFASCNFromCert(&cert, fascn, &fascnSz), 0);
+    XFREE(fascn, DYNAMIC_TYPE_TMP_BUFFER, NULL);
+
+    AssertIntEQ(wc_GetUUIDFromCert(&cert, NULL, &uuidSz), LENGTH_ONLY_E);
+    uuid = (byte*)XMALLOC(uuidSz, DYNAMIC_TYPE_TMP_BUFFER, NULL);
```
Comment (dgarske) [inline tests/api.c:2291]: Should be `uuid = (byte*)XMALLOC(uuidSz, NULL, DYNAMIC_TYPE_TMP_BUFFER);`

File: tests/api.c [hunk 4]
```diff
@@ -2260,6 +2260,44 @@ static void test_wolfSSL_CertManagerNameConstraint5(void)
...
+    bytes = (int)XFREAD(buf, 1, sizeof(buf), f);
+    XFCLOSE(f);
+
+    wc_InitDecodedCert(&cert, buf, bytes, NULL);
+    AssertIntEQ(wc_ParseCert(&cert, CERT_TYPE, 0, NULL), 0);
+    AssertIntEQ(wc_GetFASCNFromCert(&cert, NULL, &fascnSz), LENGTH_ONLY_E) ;
+    fascn = (byte*)XMALLOC(fascnSz, DYNAMIC_TYPE_TMP_BUFFER, NULL);
+    AssertNotNull(fascn);
+    AssertIntEQ(wc_GetFASCNFromCert(&cert, fascn, &fascnSz), 0);
+    XFREE(fascn, DYNAMIC_TYPE_TMP_BUFFER, NULL);
```
Comment (dgarske) [inline tests/api.c:2288]: Should be `XFREE(fascn, NULL, DYNAMIC_TYPE_TMP_BUFFER);`

File: tests/api.c [hunk 5]
```diff
@@ -2260,6 +2260,44 @@ static void test_wolfSSL_CertManagerNameConstraint5(void)
...
+
+    AssertIntEQ(wc_GetUUIDFromCert(&cert, NULL, &uuidSz), LENGTH_ONLY_E);
+    uuid = (byte*)XMALLOC(uuidSz, DYNAMIC_TYPE_TMP_BUFFER, NULL);
+    AssertNotNull(uuid);
+    AssertIntEQ(wc_GetUUIDFromCert(&cert, uuid, &uuidSz), 0);
+    XFREE(uuid, DYNAMIC_TYPE_TMP_BUFFER, NULL);
```
Comment (dgarske) [inline tests/api.c:2294]: Should be `XFREE(uuid, NULL, DYNAMIC_TYPE_TMP_BUFFER);`

**Review by dgarske — CHANGES_REQUESTED** [review]
Did you test this with `--enable-asn=template`?

File: wolfcrypt/src/asn.c [hunk 6]
```diff
@@ -80,6 +80,9 @@ ASN Options:
     extensions
  * WOLFSSL_HAVE_ISSUER_NAMES: Store pointers to issuer name components and their
     lengths and encodings.
+ * WOLFSSL_SUBJ_DIR_ATTR: Enable support for SubjectDirectoryAttributes
+    extension.
+ * WOLFSSL_SUBJ_INFO_ACC: Enable support for SubjectInfoAccess extension.
```
Comment (dgarske) [inline wolfcrypt/src/asn.c:85]: Can you also document `WOLFSSL_FPKI`?

File: wolfcrypt/src/asn.c [hunk 7]
```diff
@@ -14696,6 +14898,177 @@ enum {
...
+        return MEMORY_E;
+    }
+
+    XMEMCPY(cert->hwType, &input[idx], strLen);
+    cert->hwTypeSz = strLen;
+    idx += strLen;
+
+    ret = GetOctetString(input, &idx, &strLen, sz);
+    if (ret < 0)
+        return ret;
```
Comment (dgarske) [inline wolfcrypt/src/asn.c:14957]: Does this need to free `cert->hwType` or is that handled by caller?

File: wolfcrypt/src/asn.c [hunk 8]
```diff
@@ -14696,6 +14898,177 @@ enum {
...
+
+    cert->hwSerialNum = (byte*)XMALLOC(strLen + 1, cert->heap,
+                                       DYNAMIC_TYPE_X509_EXT);
+    if (cert->hwSerialNum == NULL) {
```
Comment (dgarske) [inline wolfcrypt/src/asn.c:14961]: Does this need to free `cert->hwType` or is that handled by caller?

File: wolfcrypt/src/asn.c [hunk 9]
```diff
@@ -20696,6 +21275,100 @@ int wc_GetPubKeyDerFromCert(struct DecodedCert* cert,
     return ret;
 }
 
+#ifdef WOLFSSL_FPKI
+/* Search through list for first matching alt name of the same type
+ * If 'current' is null then the search starts at the head of the list
+ * otherwise the search starts from the node after 'current' alt name.
+ * Returns 0 on success
+ */
+static DNS_entry* wc_GetAltName(struct DecodedCert* cert, int nameType,
```
Comment (dgarske) [inline wolfcrypt/src/asn.c:21284]: Seems odd to use a public named API `wc_GetAltName` for a static. Can you rename to something else?

File: wolfssl/wolfcrypt/asn.h [hunk 10]
```diff
@@ -1247,8 +1289,15 @@ struct DNS_entry {
 #if defined(OPENSSL_ALL) || defined(WOLFSSL_IP_ALT_NAME)
     char*      ipString; /* human readable form of IP address */
 #endif
+#ifdef WOLFSSL_FPKI
+    int        oidSum; /* provid oid sum for verification */
```
Comment (dgarske) [inline wolfssl/wolfcrypt/asn.h:1293]: Spelling on `provide`.

**Review by dgarske — CHANGES_REQUESTED** [review]

**Review by dgarske — APPROVED** [review]
Looks good. I think @SparkiDev should also review.

File: wolfcrypt/src/asn.c [hunk 11]
```diff
@@ -14438,22 +14589,26 @@ static void AddAltName(DecodedCert* cert, DNS_entry* dnsEntry)
 #endif
 
 #ifdef WOLFSSL_ASN_TEMPLATE
-#ifdef WOLFSSL_SEP
+#if defined(WOLFSSL_SEP) || defined(WOLFSSL_FPKI)
 /* ASN.1 template for OtherName of an X.509 certificate.
  * X.509: RFC 5280, 4.2.1.6 - OtherName (without implicit outer SEQUENCE).
  * HW Name: RFC 4108, 5 - Hardware Module Name
  * Only support HW Name where the type is a HW serial number.
```
Comment (SparkiDev) [inline wolfcrypt/src/asn.c:14596]: Add to the comment to indicate that what UPN and FASC-N are and how one of these will appear.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

**Review by SparkiDev — APPROVED** [review]

**Pull request merged**